README.md
denaro
======

**denaro**, _'money' in italian_, is a cryptocurrency written in Python.  
Maximum supply is 30.062.005.  
Maximum decimal digits count is 6.  
Blocks are generated every ~3 minutes, with a limit of 1000 transactions per block, and a limit of 2048 bytes per transaction.  
This makes possible to handle ~5 transactions per second, with a maximum block size of 2000kb.

## Setup with Docker
+ Build the base image with `make build`
+ `$ docker-compose up -d`

## Installation

Before installing denaro, you need to create the postgresql database.  
You can find the schema in [schema.sql](schema.sql).  
You have to set environmental variables for database access:
- `DENARO_DATABASE_USER`, default to `denaro`.  
- `DENARO_DATABASE_PASSWORD`, default to an empty string.  
- `DENARO_DATABASE_NAME`, default to `denaro`.  
- `DENARO_DATABASE_HOST`, default to `127.0.0.1`.  


```bash
# install postgresql
createdb denaro
```

Then install denaro.  

```bash
git clone https://github.com/denaro-coin/denaro
cd denaro
psql -d denaro -f schema.sql
pip3 install -r requirements.txt
uvicorn denaro.node.main:app --port 3006
```

Node should now sync the blockchain and start working


## Mining

denaro uses a PoW system.  

Block hash is the sha256 of raw bytes of `previous_hash, public_key, merkle_tree, timestamp, difficulty, random`, for example `387389eb614db0e3ada0af248f2f0adac12db35963f4bbdd71ef14914e8ade81dbda85e237b90aa669da00f2859e0010b0a62e0fb6e55ba6ca3ce8a961a60c64410bcfb6a038310a3bb6f1a4aaa2de1192cc10e380a774bb6f9c6ca8547f11abe3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b8556662796132007d521100`  
The derived sha256 must start with the last `difficulty` characters (hex) of the previous mined block.  
`random` must be at most 4 bytes long.  
`difficulty` can also have decimal digits (in block it is in fact `difficulty * 10`, 2 bytes long), that will restrict the `difficulty + 1`th character of the derived sha to have a limited set of values.  
```python
from math import ceil

difficulty = 6.3
decimal = difficulty % 1

charset = '0123456789abcdef'
count = ceil(16 * (1 - decimal))
allowed_characters = charset[:count]
```

Address must be present in the string in order to ensure block property.  

Blocks have a block reward that will half itself til it reaches 0.  
There will be `150000` blocks with reward `100`, and so on til `0.390625`, which will last `458732` blocks.   
The last block with a reward will be the `458733`th, with a reward of `0.3125`.  
Subsequent blocks won't have a block reward.  
Reward will be added the fees of the transactions included in the block.  
A transaction may also have no fees at all.  


CHANGELOG.md
## Next version
+ Add Docker

# 0.1.0
+ Old version

denaro/node/main.py
import ipaddress
import random
from os import environ
from typing import Union

import requests
from fastapi import FastAPI, Body
from icecream import ic
from requests import ReadTimeout
from starlette.background import BackgroundTasks
from starlette.requests import Request
from starlette.responses import JSONResponse

from denaro.helpers import timestamp, sha256, transaction_to_json
from denaro.manager import create_block, get_difficulty, Manager, get_transactions_merkle_tree, check_block_is_valid, \
    split_block_content, calculate_difficulty, clear_pending_transactions, block_to_bytes, get_transactions_merkle_tree_ordered
from denaro.node.nodes_manager import NodesManager, NodeInterface
from denaro.transactions import Transaction, CoinbaseTransaction
from denaro import Database
from denaro.constants import VERSION, ENDIAN

app = FastAPI()
db: Database = None
NodesManager.init()
nodes = NodesManager.get_nodes()
started = False
self_url = None

print = ic


def ip_is_local(ip: str) -> bool:
    try:
        addr = ipaddress.ip_address(ip)
    except:
        return False
    networks = [
        '10.0.0.0/8',
        '172.16.0.0/12',
        '192.168.0.0/16',
        '0.0.0.0/8',
        '100.64.0.0/10',
        '127.0.0.0/8',
        '169.254.0.0/16',
        '192.0.0.0/24',
        '192.0.2.0/24',
        '192.88.99.0/24',
        '198.18.0.0/15',
        '198.51.100.0/24',
        '203.0.113.0/24',
        '224.0.0.0/4',
        '233.252.0.0/24',
        '240.0.0.0/4',
        '255.255.255.255/32'
    ]
    for network in networks:
        if addr in ipaddress.ip_network(network):
            return True
    return False


def propagate(path: str, args: dict, ignore=None):
    global self_url
    nodes = NodesManager.get_nodes()
    print(args)
    for node_url in random.choices(nodes, k=5) if len(nodes) > 5 else nodes:
        _node_url = node_url
        node_url = node_url.strip('/')
        node_base_url = node_url.replace('http://', '', 1).replace('https://', '', 1)
        if (
            self_url and node_base_url == self_url.replace('http://', '', 1).replace('https://', '', 1)
        ) or (
            ignore and node_base_url == ignore.replace('http://', '', 1).replace('https://', '', 1)
        ):
            continue
        try:
            if path == 'push_block':
                r = requests.post(f'{node_url}/{path}', json=args, timeout=20, headers={'Sender-Node': self_url})
            else:
                r = requests.get(f'{node_url}/{path}', args, timeout=5, headers={'Sender-Node': self_url})
            print('node response: ', r.json())
        except Exception as e:
            print(e)
            if not isinstance(e, ReadTimeout):
                NodesManager.get_nodes().remove(_node_url)
            NodesManager.sync()


async def create_blocks(blocks: list):
    _, last_block = await calculate_difficulty()
    last_block['id'] = last_block['id'] if last_block != {} else 0
    last_block['hash'] = last_block['hash'] if 'hash' in last_block else (30_06_2005).to_bytes(32, ENDIAN).hex()
    i = last_block['id'] + 1
    for block_info in blocks:
        block = block_info['block']
        txs_hex = block_info['transactions']
        txs = merkle_tree_txs = [await Transaction.from_hex(tx) for tx in txs_hex]
        for tx in txs:
            if isinstance(tx, CoinbaseTransaction):
                txs.remove(tx)
                break
        block['merkle_tree'] = get_transactions_merkle_tree(txs) if i > 22500 else get_transactions_merkle_tree_ordered(
            txs)
        block_content = block_to_bytes(last_block['hash'], block)

        if i <= 22500:
            from itertools import permutations
            for l in permutations(merkle_tree_txs):
                txs = list(l)
                block['merkle_tree'] = get_transactions_merkle_tree_ordered(txs)
                block_content = block_to_bytes(last_block['hash'], block)
                if sha256(block_content) == block['hash']:
                    break
        assert i == block['id']
        if not await create_block(block_content.hex(), txs, last_block):
            return False
        last_block = block
        i += 1
    return True


async def _sync_blockchain(node_url: str = None):
    print('sync blockchain')
    if node_url is None:
        nodes = NodesManager.get_nodes()
        if not nodes:
            return
        node_url = nodes[0]
    node_url = node_url.strip('/')
    _, last_block = await calculate_difficulty()
    i = await db.get_next_block_id()
    node_interface = NodeInterface(node_url)
    if last_block != {}:
        remote_last_block = node_interface.get_block(i-1)['block']
        local_cache = None
        print(remote_last_block['hash'])
        if remote_last_block['hash'] != last_block['hash']:
            offset, limit = i - 500, 500
            remote_blocks = node_interface.get_blocks(i-500, 500)
            local_blocks = await db.get_blocks(offset, limit)
            local_blocks.reverse()
            remote_blocks.reverse()
            print(len(remote_blocks), len(local_blocks))
            if len(local_blocks) > len(remote_blocks):
                return
            for n, local_block in enumerate(local_blocks):
                if local_block['block']['hash'] == remote_blocks[n]['block']['hash']:
                    print(local_block, remote_blocks[n])
                    local_cache = local_blocks[:n]
                    local_cache.reverse()
                    last_common_block = i = local_block['block']['id']
                    [await db.add_pending_transaction(await Transaction.from_hex(tx)) for tx in local_block['transactions']]
                    await db.delete_blocks(last_common_block)
                    print([c['block']['id'] for c in local_cache])
                    break

    #return
    limit = 1000
    while True:
        i = await db.get_next_block_id()
        try:
            blocks = node_interface.get_blocks(i, limit)
        except Exception as e:
            print(e)
            #NodesManager.get_nodes().remove(node_url)
            NodesManager.sync()
            break
        if not blocks:
            print('syncing complete')
            return
        try:
            assert await create_blocks(blocks)
        except Exception as e:
            print(e)
            if local_cache is not None:
                await db.delete_blocks(last_common_block)
                await create_blocks(local_cache)
            return


async def sync_blockchain(node_url: str = None):
    try:
        return await _sync_blockchain(node_url)
    except Exception as e:
        print(e)
        pass


@app.on_event("startup")
async def startup():
    global db
    db = await Database.create(
        user=environ.get('DENARO_DATABASE_USER', 'denaro'),
        password=environ.get('DENARO_DATABASE_PASSWORD', ''),
        database=environ.get('DENARO_DATABASE_NAME', 'denaro'),
        host=environ.get('DENARO_DATABASE_HOST', None)
    )


@app.get("/")
def read_root():
    return {"version": VERSION, "timestamp": timestamp()}


@app.middleware("http")
async def middleware(request: Request, call_next):
    global started, self_url
    nodes = NodesManager.get_nodes()
    hostname = request.base_url.hostname

    if 'Sender-Node' in request.headers:
        NodesManager.add_node(request.headers['Sender-Node'])

    if nodes and not started or (ip_is_local(hostname) or hostname == 'localhost'):
        if not started: await sync_blockchain()
        try:
            node_url = nodes[0]
            #requests.get(f'{node_url}/add_node', {'url': })
            r = requests.get(f'{node_url}/get_nodes')
            j = r.json()
            nodes.extend(j['result'])
            NodesManager.sync()
        except:
            pass

        if not (ip_is_local(hostname) or hostname == 'localhost'):
            started = True

            self_url = str(request.base_url).strip('/')
            try:
                nodes.remove(self_url)
            except ValueError:
                pass
            try:
                nodes.remove(self_url.replace("http://", "https://"))
            except ValueError:
                pass

            NodesManager.sync()

            try:
                propagate('add_node', {'url': self_url})
            except:
                pass
    try:
        response = await call_next(request)
        response.headers['Access-Control-Allow-Origin'] = '*'
        return response
    except:
        raise
        return {'ok': False, 'error': 'Internal error'}


@app.exception_handler(Exception)
async def exception_handler(request: Request, e: Exception):
    return JSONResponse(
        status_code=500,
        content={"ok": False, "error": f"Uncaught {type(e).__name__} exception"},
    )


@app.get("/push_tx")
async def push_tx(tx_hex: str, background_tasks: BackgroundTasks):
    tx = await Transaction.from_hex(tx_hex)
    if await db.add_pending_transaction(tx):
        background_tasks.add_task(propagate, 'push_tx', {'tx_hex': tx_hex})
        return {'ok': True, 'result': 'Transaction has been accepted'}
    else:
        return {'ok': False, 'error': 'Transaction has not been added'}


@app.post("/push_block")
@app.get("/push_block")
async def push_block(request: Request, background_tasks: BackgroundTasks, block_content: str = '', txs='', body=Body(False), id: int = None):
    if body:
        txs = body['txs']
        if 'block_content' in body:
            block_content = body['block_content']
    if isinstance(txs, str):
        txs = txs.split(',')
        if txs == ['']:
            txs = []
    previous_hash = split_block_content(block_content)[0]
    next_block_id = await db.get_next_block_id()
    if id is not None:
        if next_block_id < id:
            await sync_blockchain(request.headers['Sender-Node'] if 'Sender-Node' in request.headers else None)
            if await db.get_next_block_id() != id + 1:
                return {'ok': False, 'error': 'Could not sync blockchain'}
        if next_block_id > id:
            return {'ok': False, 'error': 'Too old block'}
    else:
        previous_block = await db.get_block(previous_hash)
        if previous_block is None:
            if 'Sender-Node' in request.headers:
                await sync_blockchain(request.headers['Sender-Node'])
                return {'ok': False,
                        'error': 'Previous hash not found, had to sync according to sender node, block may have been accepted'}
            else:
                return {'ok': False, 'error': 'Previous hash not found'}
        id = previous_block['id']
    try:
        final_transactions = []
        hashes = []
        for tx_hex in txs:
            if len(tx_hex) == 64:  # it's an hash
                hashes.append(tx_hex)
            else:
                final_transactions.append(await Transaction.from_hex(tx_hex))
        if hashes:
            pending_transactions = await db.get_pending_transactions_by_hash(hashes)
            if len(pending_transactions) < len(hashes):  # one or more tx not found
                if 'Sender-Node' in request.headers:
                    await sync_blockchain(request.headers['Sender-Node'])
                    return {'ok': False,
                            'error': 'Transaction hash not found, had to sync according to sender node, block may have been accepted'}
                else:
                    return {'ok': False, 'error': 'Transaction hash not found'}
            final_transactions.extend(pending_transactions)
        if not await create_block(block_content, final_transactions):
            """if (True or await check_block_is_valid(block_content)) and id >= next_block_id and (request and 'Sender-Node' in request.headers): # fixme
                _, last_block = await calculate_difficulty()
                if previous_hash != last_block['hash']:
                    sender_node = request.headers['Sender-Node']
                    #await db.delete_block(next_block_id - 1)
                    await sync_blockchain(sender_node)
                    return {'ok': False, 'error': 'Blockchain has been resynchronized according to sender node, block may have been accepted'}"""
            return {'ok': False}
        background_tasks.add_task(propagate, 'push_block', {
            'block_content': block_content,
            'txs': [tx.hex() for tx in final_transactions] if len(final_transactions) < 10 else txs,
            'id': id
        })
        return {'ok': True}
    except Exception as e:
        print(e)
        raise
        return {'ok': False}


@app.get("/sync_blockchain")
async def sync(node_url: str = None):
    await sync_blockchain(node_url)


@app.get("/get_mining_info")
async def get_mining_info(background_tasks: BackgroundTasks):
    Manager.difficulty = None
    difficulty, last_block = await get_difficulty()
    pending_transactions = await db.get_pending_transactions_limit(1000, True)
    if random.randint(0, 10 + len(pending_transactions)) == 0:
        background_tasks.add_task(clear_pending_transactions)
    return {'ok': True, 'result': {
        'difficulty': difficulty,
        'last_block': last_block,
        'pending_transactions': pending_transactions[:10],
        'pending_transactions_hashes': (sha256(tx) for tx in pending_transactions),
        'merkle_root': get_transactions_merkle_tree(pending_transactions[:10])
    }}


@app.get("/get_address_info")
async def get_address_info(address: str):
    outputs = await db.get_spendable_outputs(address)
    balance = sum(output.amount for output in outputs)
    return {'ok': True, 'result': {
        'balance': balance,
        'spendable_outputs': [{'amount': output.amount, 'tx_hash': output.tx_hash, 'index': output.index} for output in
                              outputs]
    }}


@app.get("/add_node")
async def add_node(url: str, background_tasks: BackgroundTasks):
    nodes = NodesManager.get_nodes()
    url = url.strip('/')
    if url == self_url:
        return {'ok': False, 'error': 'Recursively adding node'}
    if url in nodes:
        return {'ok': False, 'error': 'Node already present'}
    else:
        try:
            assert NodesManager.is_node_working(url)
            background_tasks.add_task(propagate, 'add_node', {'url': url}, url)
            NodesManager.add_node(url)
            return {'ok': True, 'result': 'Node added'}
        except Exception as e:
            print(e)
            return {'ok': False, 'error': 'Could not add node'}


@app.get("/get_nodes")
async def get_nodes():
    nodes = NodesManager.get_nodes()
    return {'ok': True, 'result': nodes}


@app.get("/get_pending_transactions")
async def get_pending_transactions():
    return {'ok': True, 'result': [tx.hex() for tx in await db.get_pending_transactions_limit(1000)]}


@app.get("/get_transaction")
async def get_transaction(tx_hash: str, verify: bool = False):
    tx = await db.get_transaction(tx_hash) or await db.get_pending_transaction(tx_hash)
    if tx is None:
        return {'ok': False, 'error': 'Transaction not found'}
    transaction = await transaction_to_json(tx, verify)
    return {'ok': True, 'result': transaction}


@app.get("/get_block")
async def get_block(block: str, full_transactions: bool = False):
    if block.isdecimal():
        block_info = await db.get_block_by_id(int(block))
        if block_info is not None:
            block_hash = block_info['hash']
        else:
            return {'ok': False, 'error': 'Block not found'}
    else:
        block_hash = block
        block_info = await db.get_block(block_hash)
    if block_info:
        block_info = dict(block_info)
        block_info['timestamp'] = int(block_info['timestamp'].timestamp())
        txs = await db.get_block_transactions(block_hash)
        return {'ok': True, 'result': {
            'block': block_info,
            'transactions': [tx.hex() for tx in txs],
            'full_transactions': [await transaction_to_json(tx) for tx in txs] if full_transactions else None
        }}
    else:
        return {'ok': False, 'error': 'Block not found'}


@app.get("/get_blocks")
async def get_blocks(offset: int, limit: int):
    blocks = await db.get_blocks(offset, limit)
    return {'ok': True, 'result': blocks}
